src/chainrules.jl

```diff
@@ -1,5 +1,5 @@
 # constructor
-function ChainRulesCore.rrule(::Type{<:BlockDiagonal}, blocks::Vector{V}) where {V}
+function ChainRulesCore.rrule(::Type{<:BlockDiagonal}, blocks::V) where {V<:AbstractVector}
```

Comment:
i think `AbstractVector` is fine here, but i'll rely on @willtebbutt to tell me if not

Reply:
If we exclude the tuple we won't have an `rrule` for constructors of `BlockDiagonals` with tuple blocks. Are you asking if it's ok not to have `rrule` for tuple blocks, or am I missing the point of the question?

Reply:
> we won't have an rrule for constructors of BlockDiagonals with tuple blocks

correct, that'd be a nice feature to add as a follow-up, but out-of-scope for this PR 

But my question is a different one. Before this PR, `blocks` were always a `Vector`, so we had `rrule(::Type{<:BlockDiagonal}, blocks::Vector)` whereas this PR allows them to be any `AbstractVector` (or `Tuple` but ignore that, since we're not adding that `rrule` here). So my question is, now that `blocks` can be any `AbstractVector`, are we okay with broadening the `rrule` from `rrule(::Type{<:BlockDiagonal}, blocks::Vector)`  to `rrule(::Type{<:BlockDiagonal}, blocks::AbstractVector)` ?